src/Robot/History.hs

```diff
@@ -0,0 +1,42 @@
+{-# LANGUAGE DeriveGeneric #-}
+
+module Robot.History where
+
+import Robot.AutomationData
+import Robot.TableauFoundation
+
+import GHC.Generics
+import Control.Monad
+
+import Data.Aeson (FromJSON, ToJSON)
+
+-- | History item is a data type storing the relevent information for
+-- our problem state at a given point in time
+data HistoryItem = HistoryItem { getOldTableau :: Tableau,
+                                getOldAutData :: AutData }
+    deriving (Show, Read, Generic)
+
+infix 5 :=>
+-- | The History is effectively a stack of history items.
+-- Design decision: History includes the present, so that we don't
+-- need to look in two places when doing analysis on present/history.
+-- Top of the stack should be the present.
+data History = NoHistory | History :=> HistoryItem
```

Comment:
Can we just use a Haskell list for this instead of making our own isomorphic type?